### ISSUE TITLE
How to solve Chicken & Egg problem / cyclic dependencies / update resources?

### ISSUE THREAD

**Issue by BerndWessels** [pulumi/pulumi#3021]
Hello, I am stuck with situation like this now a lot:

Create a `UserPool`
Create a `AppSync API`
Create and attach a `Policy` that grants the `UserPool Users` access to the `AppSync API`
Create a `Lambda` with the `AppSync API URL` as an `Environment Variable`

Now how to assign this `Lambda` as the `PostConfirmation Trigger` for the `UserPool` ?

So basically the problem is that sometimes there are cyclic / circular dependencies which can only be resolved by updating a recently created resource within the same stack update, I guess!?

**Comment by ellismg** [comment]
Hi @BerndWessels,

I don't think that we have a satisfying answer here, unfortunately.  I do think that this idea (being able to express a new desired state for a resource as part of an update) is something that's very interesting a bullet we'll have to bite off at some point, but we don't have plans to do so at this time.

We'll keep this issue open to track the the request.

**Comment by lukehoban** [comment]
> attach a `Policy`

Can this step not be used to break the cycle?  That is - can the policy be attached after the Lambda registered to the `PostConfirmation Trigger`?  Or does that somehow block the registration of the handler?

**Comment by BerndWessels** [comment]
@lukehoban thanks, yes, I had to restructure big chunks of my stack components to get them into the right order (now AppSync before UserPool) and do attach the necessary policy afterwards.
@ellismg there will definitely be situations where there is no workaround and we need to have a way to solve this properly.
Hope we can come up with something reasonable like an update/modify function on resource instances.

**Comment by Gerrit-K** [comment]
I just stumbled across this problem with what I believe should be a rather common use case:

Let's say there are two web apps (in my case Azure App Services): a backend and a frontend. The frontend obviously needs the backend URL in order to connect to it. The backend however needs the URL of the frontend host to successfully configure CORS. Since both settings are set _during/before_ resource creation, but the URLs are only known _after_ their creation, there is a cyclic dependency that seems impossible to break.

This is a very simple example and it was not that hard to spot, but as @BerndWessels already said there will surely be many more and also much more complex scenarios. 

Since there has been no activity here for almost a year (and in the linked issue for even more than 3 years), I dare to ask: is this still on the roadmap or even actively worked on? I can imagine that it would be beneficial for so many use cases, but also a pretty hard nut to crack.

**Comment by GeXiaoguo** [comment]
> Let's say there are two web apps (in my case Azure App Services): a backend and a frontend. The frontend obviously needs the backend URL in order to connect to it. The backend however needs the URL of the frontend host to successfully configure CORS. Since both settings are set _during/before_ resource creation, but the URLs are only known _after_ their creation, there is a cyclic dependency that seems impossible to break.
> 
S3 backed Web frontend that calls API Gateway endpoints is an ubiquitous pattern for Serverless Web apps. This is a blocking issue for this kind of apps.

**Comment by leezen** [comment]
I made an example of how to solve the "chicken and egg" w/ automation API in https://www.youtube.com/watch?v=N6Wn8dKgJ34 (code at https://github.com/pulumi/pulumitv/blob/master/modern-infrastructure-wednesday/2020-10-21/)

**Comment by zvictor** [comment]
I followed @leezen's instructions but, because of #5672, I started looking for a less disruptive solution.
Thus, I ended up keeping my `pulumi up` command as it was but followed by a `post-deploy` script execution right after.

That's the simplest solution I could find for this problem as it doesn't require a move of all the deployment code to the automation API, only the `post-deploy` part.

deploy.sh:
```bash
#!/bin/bash

pulumi up --skip-preview --yes --logtostderr -v=9 2> pulumi.log --tracing=file:./pulumi.trace
node_modules/.bin/ts-node post-deploy.ts
```

post-deploy.ts:
```typescript
import { LocalWorkspace } from '@pulumi/pulumi/x/automation'

const run = async () => {
  const stack = await LocalWorkspace.selectStack({ stackName: 'dev', workDir: __dirname })
  const output = await stack.outputs()

  if (!output.frontend) {
    throw new Error(`Missing output of application 'frontend'`)
  }

  console.log('\n\nPost deploy: Updating stack')
  const storedCorsDomains = await stack.getConfig('api-gateway:CORS_DOMAINS').catch(() => {})

  if (!storedCorsDomains || storedCorsDomains.value !== output.frontend.value.endpoint) {
    console.warn(`The api-gateway:CORS_DOMAINS config needs to be updated!`)

    await stack.setConfig('api-gateway:CORS_DOMAINS', { value: output.frontend.value.endpoint })
    const corsUpdateResult = await stack.up({ onOutput: (data) => process.stdout.write(data) })

    console.log(corsUpdateResult.outputs)
  } else {
    console.log('no updates needed =)')
  }
}

run().catch(console.error)
```

---

I faced 2 issues, however:

* I now have a dynamic value (`output.frontend.value.endpoint`) being stored in my config, which can lead to constant changes in my git history whenever that value change (think on a service using a port dynamically defined). Can't I append values on the fly when calling `stack.up` instead of storing them in the config file?
* The `stack.up` command inside the `post-deploy` script is not smart. It redeploys all my containers instead of updating only the `api-gateway` one. I assume that is being caused by [pulumi not being able to properly run a docker container using a local image](https://github.com/pulumi/pulumi-docker/issues/148), which leads it to actually [recreate the Container on every deploy](https://github.com/pulumi/pulumi-docker/issues/148#issuecomment-597860856).

**Comment by joelverhagen** [comment]
From what I can tell, the automation API is not available in C# yet (https://github.com/pulumi/pulumi/issues/5596). Therefore, I worked around the problem in a sort-of-gross way adding logic to an `Output` that has side-effects. You can isolate the side-effects to non-dry runs by checking `Deployment.Instance.IsDryRun`. This is probably ill-advised for reasons I don't understand but I figured I'd mention it.

For my case, I wanted to create an AAD app registration (via the AzureAD package) and consider it into an Azure App Service. The circular dependency is that you need the app service URL for the allowed redirect URLs on the app registration but you need the app registration client ID for the app service configuration.

Example is here:
https://github.com/joelverhagen/ExplorePackages/blob/b255c7564059b27eb22d9cd0ec2facf11a6606fd/src/ExplorePackages.Infrastructure/MyStack.cs#L101-L161

**Comment by joelverhagen** [comment]
Another approach I took was writing my own resource provider plug-in which performs the "wire-up" step, resolving the chicken and egg problem. I think it's possible that using [dynamic providers](https://www.pulumi.com/blog/dynamic-providers/) may be an easier approach but I am using C# so I can't use dynamic providers. Here's an example small resource provider plug-in that resolves a chicken and egg problem between Azure App Service and Azure AD app registrations:
https://github.com/joelverhagen/pulumi-knapcode

**Comment by b4stien** [comment]
Adding two new usecases to this "feature request", centered around a "a child is being replaced I need to replace its parent as well" case.

Case one. I have a list of EC2 Elastic IP pre-whitelisted at one of my provider. I iterate over the list and find one IP that is not already allocated each time an EC2 instance is created (aborting if no free IP is found). In this scenario the dynamic resource, let's call it `PreWhitelistedEipPool`, is a parent of `ec2.Instance` via its output `PreWhitelistedEipPool.selected_ip` given to the instance as input. If the instance is replaced I want `PreWhitelistedEipPool` to select another free IP and not return the previously returned one which is still associated to the instance being replaced, otherwise the whole "create before delete" strategy fails.

Case two. For provisioning purposes I'm trying to generate a new short-lived SSH key pair each time an EC2 instance is created. In this setup the EC2 instance depends on the key pair but if the EC2 Instance is replaced I have no way to mark the key pair as "needed to be replaced as well".

**Comment by swenzel** [comment]
It's not quite a solution but IMHO I found a good enough work around. It is similar to the approach from @leezen.
In order to break the cycle, the code basically checks if a resource that is part of the cycle already exists. If it does it uses the required value from that resource but if it doesn't it uses some temporary default value.
One still has to run `pulumi up` twice (or possibly more, depending on the complexity of the case) but it should be fairly simple to make your ci run it until there are no more changes.

Anyway here is my example from DigitalOcean (A-record requires IP, IP comes from LB which requires cert, cert requires A-record):

```typescript
import * as pulumi from "@pulumi/pulumi";
import * as ocean from "@pulumi/digitalocean";

const baseDomainStr = "example.com"
const subDomainStr = "mytest";
const region = ocean.Region.FRA1;
const dropletTag = "mytag:myvalue"


const baseDomain = ocean.Domain.get(baseDomainStr, baseDomainStr);
const loadBalancerName = `${subDomainStr}-loadbalancer`;

ocean.getLoadBalancer({name: loadBalancerName}).then(
    // LoadBalancer has already been created, use its IP
    (result) => setupLoadbalancer(result.ip),

    // LoadBalancer has not been created yet, use "0.0.0.0" as temporary value
    () => setupLoadbalancer("0.0.0.0")
)

function setupLoadbalancer(ip: string): ocean.LoadBalancer {
    let dnsRecord = new ocean.DnsRecord(
        `${subDomainStr}-a-record`,
        {
            name: `*.${subDomainStr}`,
            domain: baseDomain.name,
            type: "A",
            value: ip,
        },
    );

    let cert = new ocean.Certificate(
        `${subDomainStr}-cert`,
        {
            name: `${subDomainStr}-cert`,
            domains: [
                baseDomain.name,
                dnsRecord.fqdn
            ],
            type: ocean.CertificateType.LetsEncrypt
        }
    );

    let loadBalancer = new ocean.LoadBalancer(
        loadBalancerName,
        {
            name: loadBalancerName,
            dropletTag: dropletTag,
            region: region,
            forwardingRules: [{
                entryPort: 443,
                entryProtocol: "https",
                certificateName: cert.name,
                targetPort: 80,
                targetProtocol: "http",
                tlsPassthrough: false
            }],
            healthcheck: {
                port: 22,
                protocol: "tcp",
            },
        },
    );

    return loadBalancer;
}
```

**Comment by danielrbradley** [comment]
Write-up of the explorations into circular dependencies during our recent hack day: https://www.pulumi.com/blog/exploring-circular-dependencies/

**Comment by dblair-cet** [comment]
@danielrbradley has there been any progress on solving this internally?

**Comment by yonihod** [comment]
I am wondering if there is any progress on the matter?

**Comment by awoimbee** [comment]
One solution I'm exploring in typescript is to create a magic wrapper around a `Deferred` type (itself a wrapper around `Promise`): https://gist.github.com/awoimbee/a3a2c9d72f0b88420a97397f246aa292
It's a bit like [Option 2: Refs](https://www.pulumi.com/blog/exploring-circular-dependencies/#option-2-refs).

**Comment by danielrbradley** [comment]
Where we're at right now is stil to use the workaround discussed at the begging of the blog article - referencing your own stack and deploying twice. Any future developments to avoid the second deployment are likely to be a way off for the time being.

```typescript
const currentStack = new pulumi.StackReference(
  `acme-org/${pulumi.getProject()}/${pulumi.getStack()}`
);

const ping = new CallbackFunction("ping", {
  callback: pingPongHandler,
  environment: {
    variables: {
      OPPONENT_FN_NAME: currentStack
        .getOutput("pongName")
        .apply((pongName) => pongName ?? ""),
    },
  },
});

const pong = new CallbackFunction("pong", {
  callback: pingPongHandler,
  environment: {
    variables: {
      OPPONENT_FN_NAME: ping.name,
    },
  },
});

export const pingName = ping.name;
export const pongName = pong.name;
```

**Comment by danielrbradley** [comment]
A quick update on this issue: I've got a draft proposal for an approach which I think could be used to solve these cyclic dependencies in a method which wouldn't break the core requirements of Pulumi as a declaritive system.

This is being tracked as a separate issue:
- https://github.com/pulumi/pulumi/issues/13999

**Comment by niklasravnsborg** [comment]
I also have the need to be able to express circular dependencies: In Azure, you can create an SSL certificate for a domain only after this domain has already been bound to a WebApp (`WebAppHostNameBinding`). But you need the thumbprint of this certificate as an input for the `WebAppHostNameBinding`.

For now, I solved it with a `StackReference` and a global `Output` but a more expressive option that doesn't require two deploys would be really nice.

**Comment by imod** [comment]
@niklasravnsborg there is a specific issue for this case: https://github.com/pulumi/pulumi-azure-native/issues/578

**Comment by Mondonno** [comment]
Resolution of this issue is needed 💯 
Another example from azure-native, when I am trying to create PublicIp with reverse DNS using azure-native and adding these into Cloudflare, but to have the reverse FQDN setup I need a public ip address BEFORE I can set the reverse FQDN. I know this can be solved using stack references but I wouldn't like to run `pulumi up` twice since it kind of breaks the whole purpose of pulumi.
Error azure-native is throwing at me:
```
    error: Status=400 Code="ReverseFqdnDoesNotBelongToSubscription" Message="ReverseFqdn mail.iminti.com that PublicIPAddress mail-network-ip is trying to use does not belong to subscription 7fad2891-89dc-4e4c-9f1b-a2bc145af60a. One of the following conditions need to be met to establish ownership: 1) ReverseFqdn matches fqdn of any public ip resource under the subscription; 2) ReverseFqdn resolves to the fqdn (through CName records chain) of any public ip resource under the subcription; 3) It resolves to the ip address (through CName and A records chain) of a static public ip resource under the subscription."
```

**Comment by Mondonno** [comment]
Resolution of this issue is needed 💯 
Another example from azure-native, when I am trying to create PublicIp with reverse DNS using azure-native and adding these into Cloudflare, but to have the reverse FQDN setup I need a public ip address BEFORE I can set the reverse FQDN. I know this can be solved using stack references but I wouldn't like to run `pulumi up` twice since it kind of breaks the whole purpose of pulumi.
Error azure-native is throwing at me:
```
    error: Status=400 Code="ReverseFqdnDoesNotBelongToSubscription" Message="ReverseFqdn mail.[my_domain].com that PublicIPAddress mail-network-ip is trying to use does not belong to subscription [my_subscription_id]. One of the following conditions need to be met to establish ownership: 1) ReverseFqdn matches fqdn of any public ip resource under the subscription; 2) ReverseFqdn resolves to the fqdn (through CName records chain) of any public ip resource under the subcription; 3) It resolves to the ip address (through CName and A records chain) of a static public ip resource under the subscription."
```
